// handsontable/3rdparty/walkontable/src/overlay/index.js
"use strict";

require("core-js/modules/es.object.to-string.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/es.object.keys.js");
exports.__esModule = true;
var _exportNames = {
  BottomInlineStartCornerOverlay: true,
  BottomOverlay: true,
  InlineStartOverlay: true,
  Overlay: true,
  TopInlineStartCornerOverlay: true,
  TopOverlay: true
};
exports.TopOverlay = exports.TopInlineStartCornerOverlay = exports.Overlay = exports.InlineStartOverlay = exports.BottomOverlay = exports.BottomInlineStartCornerOverlay = void 0;
var _bottomInlineStartCorner = require("./bottomInlineStartCorner");
exports.BottomInlineStartCornerOverlay = _bottomInlineStartCorner.BottomInlineStartCornerOverlay;
var _bottom = require("./bottom");
exports.BottomOverlay = _bottom.BottomOverlay;
var _inlineStart = require("./inlineStart");
exports.InlineStartOverlay = _inlineStart.InlineStartOverlay;
var _base = require("./_base");
exports.Overlay = _base.Overlay;
var _topInlineStartCorner = require("./topInlineStartCorner");
exports.TopInlineStartCornerOverlay = _topInlineStartCorner.TopInlineStartCornerOverlay;
var _top = require("./top");
exports.TopOverlay = _top.TopOverlay;
var _constants = require("./constants");
Object.keys(_constants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _constants[key]) return;
  exports[key] = _constants[key];
});
