// hyperformula/commonjs/interpreter/plugin/LookupPlugin.js
"use strict";

exports.__esModule = true;
exports.LookupPlugin = void 0;

var _AbsoluteCellRange = require("../../AbsoluteCellRange");

var _Cell = require("../../Cell");

var _errorMessage = require("../../error-message");

var _RowSearchStrategy = require("../../Lookup/RowSearchStrategy");

var _statistics = require("../../statistics");

var _ArithmeticHelper = require("../ArithmeticHelper");

var _SimpleRangeValue = require("../SimpleRangeValue");

var _FunctionPlugin = require("./FunctionPlugin");

/**
 * @license
 * Copyright (c) 2022 Handsoncode. All rights reserved.
 */
class LookupPlugin extends _FunctionPlugin.FunctionPlugin {
  constructor() {
    super(...arguments);
    this.rowSearch = new _RowSearchStrategy.RowSearchStrategy(this.config, this.dependencyGraph);
  }
  /**
   * Corresponds to VLOOKUP(key, range, index, [sorted])
   *
   * @param ast
   * @param state
   */


  vlookup(ast, state) {
    return this.runFunction(ast.args, state, this.metadata('VLOOKUP'), (key, rangeValue, index, sorted) => {
      const range = rangeValue.range;

      if (range === undefined) {
        return new _Cell.CellError(_Cell.ErrorType.VALUE, _errorMessage.ErrorMessage.WrongType);
      }

      if (index < 1) {
        return new _Cell.CellError(_Cell.ErrorType.VALUE, _errorMessage.ErrorMessage.LessThanOne);
      }

      if (index > range.width()) {
        return new _Cell.CellError(_Cell.ErrorType.REF, _errorMessage.ErrorMessage.IndexLarge);
      }

      return this.doVlookup((0, _ArithmeticHelper.zeroIfEmpty)(key), rangeValue, index - 1, sorted);
    });
  }
  /**
   * Corresponds to HLOOKUP(key, range, index, [sorted])
   *
   * @param ast
   * @param formulaAddress
   */


  hlookup(ast, state) {
    return this.runFunction(ast.args, state, this.metadata('HLOOKUP'), (key, rangeValue, index, sorted) => {
      const range = rangeValue.range;

      if (range === undefined) {
        return new _Cell.CellError(_Cell.ErrorType.VALUE, _errorMessage.ErrorMessage.WrongType);
      }

      if (index < 1) {
        return new _Cell.CellError(_Cell.ErrorType.VALUE, _errorMessage.ErrorMessage.LessThanOne);
      }

      if (index > range.height()) {
        return new _Cell.CellError(_Cell.ErrorType.REF, _errorMessage.ErrorMessage.IndexLarge);
      }

      return this.doHlookup((0, _ArithmeticHelper.zeroIfEmpty)(key), rangeValue, index - 1, sorted);
    });
  }

  match(ast, state) {
    return this.runFunction(ast.args, state, this.metadata('MATCH'), (key, rangeValue, sorted) => {
      return this.doMatch((0, _ArithmeticHelper.zeroIfEmpty)(key), rangeValue, sorted);
    });
  }

  searchInRange(key, range, sorted, searchStrategy) {
    if (!sorted && typeof key === 'string' && this.arithmeticHelper.requiresRegex(key)) {
      return searchStrategy.advancedFind(this.arithmeticHelper.eqMatcherFunction(key), range);
    } else {
      return searchStrategy.find(key, range, sorted);
    }
  }

  doVlookup(key, rangeValue, index, sorted) {
    this.dependencyGraph.stats.start(_statistics.StatType.VLOOKUP);
    const range = rangeValue.range;
    let searchedRange;

    if (range === undefined) {
      searchedRange = _SimpleRangeValue.SimpleRangeValue.onlyValues(rangeValue.data.map(arg => [arg[0]]));
    } else {
      searchedRange = _SimpleRangeValue.SimpleRangeValue.onlyRange(_AbsoluteCellRange.AbsoluteCellRange.spanFrom(range.start, 1, range.height()), this.dependencyGraph);
    }

    const rowIndex = this.searchInRange(key, searchedRange, sorted, this.columnSearch);
    this.dependencyGraph.stats.end(_statistics.StatType.VLOOKUP);

    if (rowIndex === -1) {
      return new _Cell.CellError(_Cell.ErrorType.NA, _errorMessage.ErrorMessage.ValueNotFound);
    }

    let value;

    if (range === undefined) {
      value = rangeValue.data[rowIndex][index];
    } else {
      const address = (0, _Cell.simpleCellAddress)(range.sheet, range.start.col + index, range.start.row + rowIndex);
      value = this.dependencyGraph.getCellValue(address);
    }

    if (value instanceof _SimpleRangeValue.SimpleRangeValue) {
      return new _Cell.CellError(_Cell.ErrorType.VALUE, _errorMessage.ErrorMessage.WrongType);
    }

    return value;
  }

  doHlookup(key, rangeValue, index, sorted) {
    const range = rangeValue.range;
    let searchedRange;

    if (range === undefined) {
      searchedRange = _SimpleRangeValue.SimpleRangeValue.onlyValues([rangeValue.data[0]]);
    } else {
      searchedRange = _SimpleRangeValue.SimpleRangeValue.onlyRange(_AbsoluteCellRange.AbsoluteCellRange.spanFrom(range.start, range.width(), 1), this.dependencyGraph);
    }

    const colIndex = this.searchInRange(key, searchedRange, sorted, this.rowSearch);

    if (colIndex === -1) {
      return new _Cell.CellError(_Cell.ErrorType.NA, _errorMessage.ErrorMessage.ValueNotFound);
    }

    let value;

    if (range === undefined) {
      value = rangeValue.data[index][colIndex];
    } else {
      const address = (0, _Cell.simpleCellAddress)(range.sheet, range.start.col + colIndex, range.start.row + index);
      value = this.dependencyGraph.getCellValue(address);
    }

    if (value instanceof _SimpleRangeValue.SimpleRangeValue) {
      return new _Cell.CellError(_Cell.ErrorType.VALUE, _errorMessage.ErrorMessage.WrongType);
    }

    return value;
  }

  doMatch(key, rangeValue, sorted) {
    if (rangeValue.width() > 1 && rangeValue.height() > 1) {
      return new _Cell.CellError(_Cell.ErrorType.NA);
    }

    if (rangeValue.width() === 1) {
      const index = this.columnSearch.find(key, rangeValue, sorted !== 0);

      if (index === -1) {
        return new _Cell.CellError(_Cell.ErrorType.NA, _errorMessage.ErrorMessage.ValueNotFound);
      }

      return index + 1;
    } else {
      const index = this.rowSearch.find(key, rangeValue, sorted !== 0);

      if (index === -1) {
        return new _Cell.CellError(_Cell.ErrorType.NA, _errorMessage.ErrorMessage.ValueNotFound);
      }

      return index + 1;
    }
  }

}

exports.LookupPlugin = LookupPlugin;
LookupPlugin.implementedFunctions = {
  'VLOOKUP': {
    method: 'vlookup',
    parameters: [{
      argumentType: _FunctionPlugin.ArgumentTypes.NOERROR
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.RANGE
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.NUMBER
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.BOOLEAN,
      defaultValue: true
    }]
  },
  'HLOOKUP': {
    method: 'hlookup',
    parameters: [{
      argumentType: _FunctionPlugin.ArgumentTypes.NOERROR
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.RANGE
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.NUMBER
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.BOOLEAN,
      defaultValue: true
    }]
  },
  'MATCH': {
    method: 'match',
    parameters: [{
      argumentType: _FunctionPlugin.ArgumentTypes.NOERROR
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.RANGE
    }, {
      argumentType: _FunctionPlugin.ArgumentTypes.NUMBER,
      defaultValue: 1
    }]
  }
};
